### PR TITLE
Adding the ability to manually flush to segment generators.

### PR DESCRIPTION
**Pull request by chongzluong**
## Overview

We're adding the ability to manually flush through the Python Client.

A flush in this context will be called through `flush()` and result in a generator function that, when called, iterates over all audio generated for transcripts submitted since the last time (if ever) a `flush()` was called.

In other words, if you do the following:
```
ws = await client.tts.websocket()
ctx = ws.context()
await ctx.send(..., transcript=transcript_1, ...)
await ctx.send(..., transcript=transcript_2, ...)
receiver_1 = await ctx.flush()
await ctx.send(..., transcript=transcript_3, ...)
receiver_2 = await ctx.flush()
```
Then iterating over the receivers will get you the following audio:
```
async for output in receiver_1():
  # Audio Chunks for Transcript 1 and 2

async for output in receiver_2():
  # Audio Chunks for Transcript 3
```

Previously Cartesia TTS was solely multiple inputs to a single output receiver, at least with regards to [continuations](https://docs.cartesia.ai/build-with-sonic/capability-guides/stream-inputs-using-continuations). This is because we perform smart splitting & merging on our end to optimize model performance.

However, many of the existing TTS architectures have multiple inputs to multiple output generators, so we should support this flexibility to accommodate for integrating with other providers with as minimal friction as possible.

If you call `flush()` it'll be because you've performed chunking yourself (ideally the transcript or transcript(s) submitted are a sentence).

## Implementation

Currently this is implemented using a similar mechanism to how we wrapped multiplexing over a single websocket. We performed this by iterating over the receiver asynchronously in the background, and separating chunks into Async Queues mapped by their Context ID. In this case, we separate chunks into Lists of Async Queues, where their Flush ID represents the index of the List where the queue we're populating is. When `flush()` is called, we append a new Asynchronous Queue to the list for that Context ID.

multiplexing_flushing_python

In a previous iteration of this PR I was doing this non-deterministically using Async Polling from the N queue and the N+1 queue. Instead we introduced a notion of `flush_done` event fired after manual flushes prior to incrementing the `flush_id`. This allows us to deterministically indicate that the generator function has completed.

## Testing

Tested on a local API version - Prod already added the manual flush capabilities but my local branch has the changes that incorporate the `flush_done` event. Confirmed that given 3 transcripts, I can create the 3 generator functions and iterate over each for the distinct audio generations.

Also added a unit test that creates 3 generators from 3 transcript + flushes and iterates over each of them. This currently passes in the staging deployment, this PR is blocked until the flush done emission is deployed in production.

### CODE REVIEW

File: cartesia/_async_websocket.py [hunk 1]
```diff
@@ -105,6 +108,95 @@ async def no_more_inputs(self) -> None:
             context_id=self._context_id,
             continue_=False,
         )
+    
+    async def flush(self) -> Callable[[], AsyncGenerator[Dict[str, Any], None]]:
+        """Trigger a manual flush for the current context's generation. This method returns a generation that yields the audio prior to the flush."""
+        await self.send(
+            model_id=DEFAULT_MODEL_ID,
+            transcript="",
+            output_format=TTS.get_output_format("raw_pcm_f32le_44100"),
```
Comment (nimz) [inline cartesia/_async_websocket.py]: should we define the default output format somewhere instead of passing to `self.send` each time as a 'magic constant'

File: cartesia/_async_websocket.py [hunk 2]
```diff
@@ -105,6 +108,95 @@ async def no_more_inputs(self) -> None:
...
+                    # 2. Next queue to get a message while current queue is empty
+                    # 3. Timeout to occur (only if next queue is empty)
+                    # Print the length of each queue
+                    done, pending = await asyncio.wait(
+                        tasks,
+                        return_when=asyncio.FIRST_COMPLETED,
+                        timeout=self.timeout
+                    )
+
+                    completed_or_cancelled_tasks = [completed_task for completed_task in done]
```
Comment (nimz) [inline cartesia/_async_websocket.py]: nit: `cancelled` -> `canceled` (all instances) [`cancelled` is British spelling]

File: cartesia/_async_websocket.py [hunk 3]
```diff
@@ -105,6 +108,95 @@ async def no_more_inputs(self) -> None:
...
+                    
+                    # If next queue exists, create a task to check it
+                    next_queue_task = asyncio.create_task(self._websocket._get_message(self._context_id, timeout=self.timeout, flush_id=flush_id + 1))
+                    tasks.append(next_queue_task)
+
+                    # Wait for either:
+                    # 1. Current queue to get a message
+                    # 2. Next queue to get a message while current queue is empty
+                    # 3. Timeout to occur (only if next queue is empty)
+                    # Print the length of each queue
```
Comment (nimz) [inline cartesia/_async_websocket.py]: where is this print done?

File: cartesia/_async_websocket.py (hunk 1 above)
Comment (chongzluong) [reply cartesia/_async_websocket.py]: this makes almost too much sense

File: cartesia/_async_websocket.py (hunk 2 above)
Comment (chongzluong) [reply cartesia/_async_websocket.py]: I'm a big fan of the Queen's English

File: cartesia/_async_websocket.py (hunk 3 above)
Comment (chongzluong) [reply cartesia/_async_websocket.py]: op i missed this when I was removing some debugging logs mb

**Comment by chongzluong** [comment]
Tests will pass when production deployment is updated (I ran the tests on staging)

File: cartesia/version.py
Comment (kbrgl) [inline cartesia/version.py:1]: We now use use bumpversion.py to bump the version instead of doing it manually, that'll also update the pyproject.toml for ya

File: cartesia/_async_websocket.py [hunk 4]
```diff
@@ -282,18 +322,25 @@ async def _process_responses(self):
...
             raise e
 
-    async def _get_message(self, context_id: str, timeout: float) -> Dict[str, Any]:
+    async def _get_message(
+        self, context_id: str, timeout: float, flush_id: Optional[int] = -1
+    ) -> Dict[str, Any]:
         if context_id not in self._context_queues:
             raise ValueError(f"Context ID {context_id} not found.")
-        return await asyncio.wait_for(self._context_queues[context_id].get(), timeout=timeout)
+        return await asyncio.wait_for(
```
Comment (kbrgl) [inline cartesia/_async_websocket.py:341]: Does this need a check for `flush_id not in self._context_queues[context_id]`?

File: cartesia/_async_websocket.py [hunk 5]
```diff
@@ -213,7 +253,7 @@ async def close(self):
             except asyncio.CancelledError:
                 pass
             except TypeError as e:
-                # Ignore the error if the task is already cancelled
+                # Ignore the error if the task is already canceled
```
Comment (kbrgl) [inline cartesia/_async_websocket.py]: 🇺🇸🇺🇸🇺🇸

**Review by kbrgl — APPROVED** [review]

File: cartesia/version.py
Comment (chongzluong) [reply cartesia/version.py:1]: oh gotcha mbmb

File: cartesia/_async_websocket.py (hunk 4 above)
Comment (chongzluong) [reply cartesia/_async_websocket.py:341]: This makes too much sense

File: cartesia/_async_websocket.py [hunk 6]
```diff
@@ -100,12 +103,49 @@ async def no_more_inputs(self) -> None:
...
             transcript="",
-            output_format=TTS.get_output_format("raw_pcm_f32le_44100"),
+            output_format=TTS.get_output_format(DEFAULT_OUTPUT_FORMAT),
             voice_embedding=DEFAULT_VOICE_EMBEDDING,  # Default voice embedding since it's a required input for now.
             context_id=self._context_id,
             continue_=False,
         )
 
+    async def flush(self) -> Callable[[], AsyncGenerator[Dict[str, Any], None]]:
+        """Trigger a manual flush for the current context's generation. This method returns a generation that yields the audio prior to the flush."""
```
Comment (ad12) [inline cartesia/_async_websocket.py]:
```suggestion
        """Trigger a manual flush for the current context's generation. This method returns a generator that yields the audio prior to the flush."""
```

File: cartesia/_async_websocket.py (hunk 5 above)
Comment (ad12) [inline cartesia/_async_websocket.py]:
```suggestion
                # Ignore the error if the task is already canceled.
```

File: cartesia/_async_websocket.py [hunk 7]
```diff
@@ -282,18 +322,27 @@ async def _process_responses(self):
         try:
             while True:
                 response = await self.websocket.receive_json()
+                if response.get("flush_done"):
+                    print(f"Flush done received for flush ID {response.get('flush_id')}")
```
Comment (ad12) [inline cartesia/_async_websocket.py]: remove print statement

File: cartesia/version.py
Comment (ad12) [inline cartesia/version.py:1]: version bumps should always be a separate independent pr

File: cartesia/_async_websocket.py [hunk 8]
```diff
@@ -100,12 +103,49 @@ async def no_more_inputs(self) -> None:
...
             model_id=DEFAULT_MODEL_ID,
             transcript="",
-            output_format=TTS.get_output_format("raw_pcm_f32le_44100"),
+            output_format=TTS.get_output_format(DEFAULT_OUTPUT_FORMAT),
             voice_embedding=DEFAULT_VOICE_EMBEDDING,  # Default voice embedding since it's a required input for now.
             context_id=self._context_id,
             continue_=False,
         )
 
+    async def flush(self) -> Callable[[], AsyncGenerator[Dict[str, Any], None]]:
```
Comment (ad12) [inline cartesia/_async_websocket.py:112]: in this case, if one flushes, they have to use the `generator` returned by `flush()`. they can't use the standard `receive` right?

Comment (chongzluong) [reply cartesia/_async_websocket.py:112]: They technically can, it's just not advised

Comment (chongzluong) [reply cartesia/_async_websocket.py:112]: In the context of the implementation we've chosen, calling `receive()` will just get u the audio from the latest flush ID

File: cartesia/version.py
Comment (chongzluong) [reply cartesia/version.py:1]: gotcha
